RFC: added native ami versioning design doc (#3356)
aws/karpenter: designs/native-ami-versioning.md

# Native AMI Versioning

> Karpenter uses the latest EKS Optimized AMI releases as defaults in their AMI Requirements. With [Drift](https://karpenter.sh/preview/concepts/deprovisioning/#drift) implemented, when a new EKS Optimized AMI is released users who have enabled drift could see their fleet rolled out to the newly released AMI without warning.

Reviewer: The sentence structure is a bit confusing here.  I think it's worth calling out the different "EKS Optimized" AMIs in this intro section too since that can be another point of confusion on exactly what we're talking about here. 

```suggestion
Karpenter uses the latest EKS Optimized AMI releases for Bottlerocket, Amazon Linux 2, and Ubuntu as the defaults for AMI Requirements. With [Drift](https://karpenter.sh/preview/concepts/deprovisioning/#drift) implemented, when a new EKS Optimized AMI is released, users who have enabled drift could see the newly released AMIs rolled out to their fleet without warning.
```

## What could happen when a new EKS Optimized AMI is released?

When a new EKS Optimized AMI is released, Karpenter could begin deprovisioning and replacing nodes. If a user wants to revert back to an older release, they need to find the corresponding AMI IDs and change their AWSNodeTemplate to pin those values by setting their AWSNodeTemplate to the following:

```
spec:
    amiSelector:
        aws-ids: "ami-123,ami-456"
```

Once the user wants to return to the latest release, they’ll have to restore their configurations to the previous default as the following.

```
spec:
    amiFamily: Ubuntu # Or Bottlerocket or AL2
```

While this isn’t a difficult process, it isn't viable to require users to jump through these hoops. They must be aware of new EKS Optimized AMI releases, understand SSM aliases to fetch the AMI IDs, and modify their AWSNodeTemplate values in their CI/CD pipelines multiple times. Some users have [requested these improvements](https://github.com/aws/karpenter/issues/1495). In order to improve the experience for all users, we’ll introduce (1) Bake Time and (2) Release Version Pinning improvements.

## Bake Time

> ### 🔑 Introduce a minimumAgeDays field in the amiSelector.

Reviewer: Days are not necessarily all the same length of time, would hours be better here and then we could use a go duration? 

I'm not sure minimumAge is a good concept for the feature we're trying to create here. ttlAfterCreated seems nicer since it matches the other ttl semantic, with the obvious downside that "Created" is somewhat ambiguous Is it SSM parameter created or AMI created? I think it needs to the SSM parameters changed since it's the standard practice to build the AMI early and then flip the SSM parameter when actually released.

Author: Sorry, this is from a previous version of the doc. I had changed all occurrences to `minimumAgeDuration` except this one 🤦 . Does your comment still hold after it being changed to `minimumAgeDuration`?

Author: Discussed offline, the SSM Parameter is updated in a varied amount of time since the image creation. Because of that we can special case this `minimumAgeDuration` to work based off the SSM Parameter creation.

> While this hasn’t been asked for explicitly, bake time is a common concept in phasing in new changes. Whether a user has a mandated bake time for production or developers have an estimated time to validate workloads in before rolling out to prod, bake time provides a sliding window of leeway. Users can be cautious and set a high value, or leave out the setting for no bake time requirements.

Reviewer: ```suggestion
While this hasn’t been asked for explicitly, bake time is a common concept in phasing in new changes. Whether a user has a mandated bake time for production or developers have an estimated time to validate workloads before rolling out to prod, bake time provides a sliding window of leeway. Users can be cautious and set a high value, or leave out the setting for no bake time requirements.
```

> While this hasn’t been asked for explicitly, bake time is a common concept in phasing in new changes. Whether a user has a mandated bake time for production or developers have an estimated time to validate workloads before rolling out to prod, bake time provides a sliding window of leeway. Users can be cautious and set a high value, or leave out the setting for no bake time requirements.

Reviewer 2: I'm cautious about whether or not this is a good idea.

```
apiVersion: karpenter.k8s.aws/v1alpha1
kind: AWSNodeTemplate
metadata:
  name: default
spec:
  amiSelector:

>     minimumAgeDuration: 10

Reviewer: I thought we had talked about duration! :) Was the above just outdated?

Author: Yes! Sorry about that.

```

### How `minimumAgeDuration` interacts with the other methods of AMI selection

#### *EKS Optimized AMI (no other AMI Selector field)*

1. The newest EKS Optimized AMI that fits the `minimumAgeDuration` selector will be chosen.

> 2. An AMI will be considered eligible at the time and day that it was created as shown in aws ec2 describe-images.

Reviewer: I think this is probably not right. See above comment.

> 1. Karpenter will compare `minimumAgeDuration` to when the SSM Parameter was last modified instead of using the Image Creation Date that's returned in `aws ec2 describe-images`. This is because the SSM Parameter is updated independently of image creation.

Reviewer 3: This might be a bit tricky, as the corresponding SSM Parameters have no value history and only the most recent value is available. Let's assume `minimumAgeDuration: 15d` and that there is a new AMI published every `7d`, then the bake time requirement will never be met and the nodes would be stuck on the current AMI version in use.

#### *Arbitrary tags*

1. An AMI tagged and discovered via the amiSelector tags will not be chosen unless it fits the `minimumAgeDuration` filter.
2. This has the downside of resulting in no AMIs found if users are too restrictive with tags and the `minimumAgeDuration` filter. The upside is that users will know any AMI created before x days ago will not be included.

#### *AMI IDs*

1. When using AMI IDs, other tags specified are not respected. To keep explicit AMI ID usage as an override, using `minimumAgeDuration` and AMI IDs will result in `minimumAgeDuration` not being respected.

### DriftTTL in the ProvisionerSpec

> Karpenter could include a drift TTL in the Provisioner. Nodes would have to be drifted for the TTL's duration to be deprovisioned. This can implement bake time, but will also apply for all future methods of Drift, such as zonal and instance permission requirements. Since other Drift methods may need more immediate actions when it occurs, setting a Drift TTL may not be granular as needed. This

Reviewer: ```suggestion
Karpenter could include a drift TTL in the Provisioner. Nodes would have to be drifted for the TTL's duration to be deprovisioned. This can implement bake time, but will also apply for all future methods of Drift, such as zonal and instance permission requirements. Since other Drift methods may need more immediate actions when it occurs, setting a Drift TTL may not be granular as needed.
```

### 🔑 Suggestion

Implement Bake Time with `minimumAgeDuration`. Users can take advantage of it with EKS Optimized AMIs generally to follow requirements on AMI bake time. This applies only to AMIs, rather than a Drift TTL which would apply broadly to every provisioning requirement. Users can rely on `minimumAgeDuration` to use it as a safeguard against unexpected AMI upgrades, creating more confidence in relying on Karpenter to make the right decisions.

## Release Pinning

*🔑 Expand amiFamily to allow a pinned version for EKS Optimized AMIs.*

Add an option to specify AMI Release Version with AMI Family. A user can pin known AMI Versions, and declaratively upgrade to a different AMI when ready. This could prevent automatic updates in the event of a newer EKS Optimized AMI. Some users [have requested this](https://github.com/aws/karpenter/issues/1495) with 13 up-votes.

### *Option 1*

Use an AMI Family with an appended magic string. This allows users who only know of version strings to modify AWSNodeTemplates easily to their liking.

```
apiVersion: karpenter.k8s.aws/v1alpha1
kind: AWSNodeTemplate
metadata:
  name: default
spec:

>   amiFamily: AL2-v20220824 # No version suffix defaults to latest

Reviewer 2: While this is a cool semantic that maps to container image semantics (e.g. amiFamily: AL2:v20220824), I think we're mixing up concepts here.

amiFamily is currently a runtime contract that tells Karpenter what to expect for the user data, ebs volumes, etc. Users can specify amiFamily as well as specific amiIDs and this semantic would enable them to specify both simultaneously. 

Does it make sense to support pinning through amiID instead?

Reviewer 4: > Does it make sense to support pinning through amiID instead?
as amiIDs are different in different regions (and for different architectures, gpu / non-gpu setups, etc), if we allow to pin only using amiID, we force users to perform tedious and error-prone lookups (while we could allow to specify some selected optional, potentially even amiFamily-specific constraints)

allowing to specify a version could be a great addition to the `minimumAgeDuration` bake time
(not saying that specifying these optional constraints in the `amiFamily` field itself is the best option)

```

This benefits current users of EKS Optimized AMIs. It extends cleanly into defaults, and allows users an easy override that follows the same [versioning semantics](#versioning-semantics) used for each of the AMI Families, which users would be aware of.

The downside is users that use custom Bootstrapping logic but do not use EKS Optimized AMIs may find this confusing if they find it in documentation or code. Confusion can be mitigated with good documentation as it’s only referring to EKS Optimized AMIs.

### *Option 2*

Include amiVersion as an additional filter *outside* of the amiSelector.

This may be more intuitive to users that are using EKS Optimized AMIs, but has no value add for those who aren’t. This may be confusing to users looking to extend the amiVersion field into other methods of custom AMI selection, especially since it’s a top level field. One alternative is to make this mutually exclusive with AMISelector so that there’s no room for confusion.

```
apiVersion: karpenter.k8s.aws/v1alpha1
kind: AWSNodeTemplate
metadata:
  name: default
spec:
  amiFamily: Bottlerocket

>   amiVersion: v1.9.1 # Defaults to latest, requires AMIFamily if specified

Reviewer 5: I love this if it can be part of the AMISelector API change - this would meet our needs perfectly

>   amiVersion: v1.9.1 # Defaults to latest, requires AMIFamily if specified

Reviewer 3: I am also in favor of this, as we could further extend `amiVersion` to support [semver version ranges](https://devhints.io/semver). Some examples: 
- `1.9.1` pin to exactly this version
- `^1.9` allow for any bug-fix release upgrades
- `^1` allow for minor release upgrades
-  ...

Author: This would be a cool extension! For specifically the EKS Optimized AMIs, this would only work for Bottlerocket. My only qualm with this option is that we'd need to add some way to plug this into custom AMIs for it to be useful for all users.

Reviewer 4: also `minimumAgeDuration` bake time should then probably be ignored only for a specific version (similarly like when amiID is specified) but not for a range

```

### 🔑 Suggestion

> Implement only Option 1. This keeps configuration within the amiFamily field, does not add new fields to be managed, and works well for users who are aware of EKS Optimized semantics, which is the audience for this feature.

Reviewer: This is easily moved too if we decide to move the amiFamily logic into the amiSelector or some other richer API like in the suggestion [here](https://github.com/aws/karpenter/pull/3204/files#diff-e08ed94660726c8fa676b3fdf3d9c1640b4207419768252f2f19f07825d2fe85).

## Appendix

### EKS Optimized AMI Releases

> EKS Optimized AMIs are not released on a regular cadence, and do not all use the same versioning semantic.

Reviewer 2: Did we explore changing this?
